### DISCUSSION TITLE
Apply changes from library function signature to program

### DISCUSSION THREAD

**Discussion by j3s5e4** [NationalSecurityAgency/ghidra#5052] [General]
How do I apply a change in a library function signature to a program using said library?

I have tried:
1. Create program archive
2. Run "Capture Function Data Types" in the library by right clicking the new program archive. This creates a folder with all function signatures.
3. Run "Apply Function Data Types" in the program by right clicking the new program archive. This does nothing.

When I right click a signature in the archive and select "Find Uses Of", it shows no results. If I drag and drop the signature to a function, it applies the signature and shows in the "Find Uses Of" window. However, when I update the library signature and try to apply the changes, it still does not update the function in the program.

**Comment by emteere** [comment]
I'm not quite sure the steps you are doing above, because it does do what you suggest it should do when I tried something that is equivalent.

- Open Program A
- Create a function called foo(int, int) in program A
- open Library Program B, create a function foo(float, float) in Library Program B
- Create a File Data Type Archive C
- While still in Library Program B, Run "Capture Function Data Types"
- Go back to Program A
- Run "Apply Function Data Types" from File Data Type Archive C
- See that Program A foo(int, int) is changed to foo(float, float)

When you use "Apply Function Data Types", the only requirement is that the function definition name match the function label name exactly.  It should overwrite the signature if there is a function with that same exact name in the program.  If the function signatures are in a namespace, I believe the namespace must then match in both places.

Once a function definition is applied from an archive to a function in a program, there is no longer any relationship with the function definition, and the function signature defined at an address in the program.  This is not true for other data types like structures, there is a direct correspondence.  If you change a structure in the DTManager, it will change in the program where ever it is defined. That will most likely change in the future function signatures.

Given the above, if you use "Find Uses Of" it won't find the uses of the function definition, unless is is a pointer to that definition, which would be a data type other than a function definition (it would be a pointer).

However, if you add data types from an file archive to a program there will be a relationship of the data types from the file archive to the program data types that came from the archive.  You will then get an "Update Data Types From->" choice in your program.  Unfortunately this does not apply to function definitions that you apply directly to the program using "Apply Function Data Types" from a File Data Type Archive.  It should in my opinion.  If you drag the function definition from the file archive to the Program Data Type Manager, then it will update that signature using "Update Data TYpes From".  But you will still have to use the Apply Function Data Types from the program to actually update any functions with that name in the program.

**Reply by j3s5e4** [reply]
It seems to work now following the steps above. I'm not sure why it didn't work before but if I have trouble again I will try to replicate it.

For overloaded functions of the same name but different signatures, is there no way to automatically apply those correctly assuming the external location for the thunks are set correctly? The captured functions default to foo.conflictN, and when I apply it, all of the functions are set to the non-confict signature.
